usersession/userd/settings.go

```diff
@@ -78,14 +94,102 @@ func allowedSetting(setting string) bool {
 }
 
 func settingWhitelisted(setting string) *dbus.Error {
+	// Trim off any subproperty from the check
+	settingTrimmed := strings.SplitN(string(setting), "/", 2)[0]
 	for _, whitelisted := range settingsWhitelist {
-		if setting == whitelisted {
+		if settingTrimmed == whitelisted {
 			return nil
 		}
 	}
 	return dbus.MakeFailedError(fmt.Errorf("cannot use setting %q: not allowed", setting))
 }
 
+func safeSnapFromSender(s *Settings, sender dbus.Sender) (string, *dbus.Error) {
+	// avoid information leak: see https://github.com/snapcore/snapd/pull/4073#discussion_r146682758
+	snap, err := snapFromSender(s.conn, sender)
+	if err != nil {
+		return "", dbus.MakeFailedError(err)
+	}
+	return snap, nil
+}
+
+func desktopFile(s *Settings, command string, setting string, desktopValue string, sender dbus.Sender) (string, *dbus.Error) {
+	snap, err := safeSnapFromSender(s, sender)
+	if err != nil {
+		return "", err
+	}
+	if err := settingWhitelisted(setting); err != nil {
+		return "", err
+	}
+
+	if !allowedSetting(desktopValue) {
+		return "", dbus.MakeFailedError(fmt.Errorf("cannot %q setting %q to value %q: value not allowed", command, setting, desktopValue))
+	}
+
+	// FIXME: this works only for desktop files
+	desktopFile := fmt.Sprintf("%s_%s", snap, desktopValue)
+	return desktopFile, nil
+}
+
+func outputDesktopFile(s *Settings, output string, sender dbus.Sender) (string, *dbus.Error) {
```

Comment:
I found the name a bit confusing and ended up jumping back and forth while reviewing the PR. Can we rename this to `desktopFileFromOutput` maybe?